rfd/0006/readme.mdx

```diff
@@ -0,0 +1,293 @@
+inefficient. While this RFD doesn't prioritize optimizing the update itself (e.g., using `ALTER VIEW`), the proposed solution provides a
+better foundation for future improvements by correctly ordering even drop/recreate steps.
+
+## Goals
+
+*   Ensure the correct order of DDL execution for dependent ClickHouse objects managed by Moose.
+*   Improve the reliability and predictability of infrastructure updates during `moose dev`.
+*   Prevent common DDL errors caused by incorrect change application order.
+*   Maintain the declarative nature of Moose infrastructure definitions for the user.
+*   Establish a robust pattern for handling SQL dependencies.
+
+## Non-Goals
+
+*   Implementing full DDL transactionality for ClickHouse changes (a ClickHouse limitation).
+*   Optimizing the update mechanism for MVs/Views beyond drop-and-recreate within *this* RFD.
+*   Managing dependencies for resources *not* managed by Moose within the ClickHouse database.
+*   Parsing SQL `SELECT` statements to automatically infer dependencies (we rely on explicit declaration).
+
+## Proposed Solution: Explicit Dependencies & Phased Ordering
+
+We propose a two-part solution: mandating explicit dependency tracking in the APIs and introducing an intermediate step for dependency analysis
+and phased, ordered execution planning.
+
+### Part 1: Explicit Dependency Tracking via APIs (Mandatory)
+
+Robust dependency management requires explicit declaration by the user. The SDK layer will provide an ergonomic way to declare these dependencies.
+
+1.  **API/SDK Changes (e.g., `ts-moose-lib`):**
+    *   When defining resources like `MaterializedView`, `View`, etc., users **must** specify *all* direct Moose resource instances they depend
+        on using a new `depends_on` property. Users will pass the *actual instances* of the Moose classes (e.g., `OlapTable`, `View` instances)
+        into this array.
+    *   **SDK Responsibility:** The Moose SDK layer (during the code execution phase that generates the JSON for the Rust CLI) will introspect
+        these provided instances. It will extract the resource type (e.g., "Table", "View", "SqlResource") and the unique ID from each instance
+        passed into the `depends_on` array.
+    *   **JSON Output:** The SDK transforms this list of instances into the structured dependency JSON format
+        `[{ resource_type: "TypeString", id: "UniqueIDString" }, ...]` which is then sent to the Rust CLI.
+    *   Example (Conceptual TS - User Code):
+        ```typescript
+        import { sourceTable1, sourceTable2, intermediateView } from './resources';
+        // Assume sourceTable1 etc. are instances of OlapTable, View...
+
+        new MaterializedView<TargetTableType>('my_mv', {
+            materializedViewName: 'my_mv_name',
+            selectStatement: `SELECT ... FROM ${sourceTable1.name} JOIN ${sourceTable2.name} ...`,
```

Comment:
the `selectStatement` can be a `sql` interpolated string. We can look for its dependencies in the interpolated values.

Reply:
Yeah but it also can be a hardcoded string. So I am thinking we start with explicit and then we are making it optional if there is string interpolation.